docs/howto/logging.txt

```diff
@@ -0,0 +1,336 @@
+:py:mod:`handler classes <logging.handlers>` provided by Python.
+
+Logging levels can also be set on the handlers (by default, they accept log
+messages of all levels). Using the example above, adding:
+
+..  code-block:: python
+    :emphasize-lines: 4
+
+    {
+        'class': 'logging.FileHandler',
+        'filename': 'general.log',
+        'level': 'DEBUG',
+    }
+
+would define a handler configuration that only accepts records of level
+``DEBUG`` and higher.
+
+Configure a logger mapping
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To send records to this handler, configure a logger mapping to use it for
+example:
+
+..  code-block:: python
+    :emphasize-lines: 3-8
+
+    LOGGING = {
+        [...]
+        'loggers': {
+            '': {
+                'level': 'DEBUG',
+                'handlers': ['file'],
+            },
+        },
+    }
+
+The mapping's name determines which log records it will process. This
+configuration (``''``) is *unnamed*. That means that it will process records
+from *all* loggers (see :ref:`naming-loggers` below on how to use the mapping
+name to determine the loggers for which it will process records).
+
+It will forward messages of levels ``DEBUG`` and higher to the handler named
+``file``.
+
+Note that a logger can forward messages to multiple handlers, so
+the relation between loggers and handlers is many-to-many.
+
+If you execute::
+
+    logger.debug('Attempting to connect to API')
+
+in your code, you will find that message in the file ``general.log`` in the
+root of the project.
+
+Configure a formatter
+~~~~~~~~~~~~~~~~~~~~~
+
+By default, the final log output contains the message part of each :class:`log
+record <logging.LogRecord>`. Use a formatter if you want to include additional
+data. First name and define your formatters - this example configures
+formatters named ``verbose`` and ``simple``:
+
+..  code-block:: python
+    :emphasize-lines: 3-12
+
+    LOGGING = {
+        [...]
+        'formatters': {
+            'verbose': {
+                'format': '{name} {levelname} {asctime} {module} {process:d} {thread:d} {message}',
+                'style': '{',
+            },
+            'simple': {
+                'format': '{levelname} {message}',
+                'style': '{',
+            },
+        },
+    }
+
+The ``style`` keyword allows you to specify ``{`` for :meth:`str.format` or
+``$`` for :class:`string.Template` formatting; the default is ``$``.
+
+See :ref:`logrecord-attributes` for the :class:`~logging.LogRecord` attributes
+you can include.
+
+To apply a formatter to a handler, add a ``formatter`` entry to the handler's
+dictionary referring to the formatter by name, for example:
+
+..  code-block:: python
+    :emphasize-lines: 5
+
+    'handlers': {
+        'file': {
+            'class': 'logging.FileHandler',
+            'filename': 'general.log',
+            'formatter': 'verbose',
+        },
+    },
```

Comment:
I'd be inclined to explain formatters first (before handlers) and then use the formatter we defined in the handler, so as to not have to cut back to handlers here. 🤔

(Order one, often, defines them in the dictConfig is `formatters`, `handers`, `loggers`, since then it reads in definition order.)

Reply:
Just to throw a spanner in the works, conversely, I'd want to read about formatters _last_. If I'm setting up logging (for the first time, or just as a refresher), the first thing I want to see is the logger call working, which requires a logger and maybe/probably a handler to the destination I want (eg: console, logstash, whatever).

It's only once I've successfully got logging configured and visibly working that I care about formatting the log records. As minor anec-data for that POV, I literally cannot recall the last time I bothered defining a custom formatter. Certainly years ago :) YMMV, naturally.

Reply:
Maybe we don't need to mention custom formatters here at all? 🤔

(Although I do like them 😀)

Reply:
This dilemma is exactly why the guide tackles them in that order. I don't think there is a really elegant solution but I will have another look.

Reply:
In the end I couldn't really find a better solution. 

* defining them all in order of dependency means defining the rather inessential formatters first
* defining them in logger > handler > formatter order would be conceptually better, but would mean cutting back twice instead of once (this would be my second-best ~~choice~~ compromise.

Reply:
OK, let's go with it. Thanks @evildmp